src/NuGet.Jobs.Catalog2Registration/Hives/HiveMerger.cs

```diff
@@ -0,0 +1,311 @@
+
+                    if (catalog.PackageIdentity.Version > pageInfo.Upper)
+                    {
+                        // The current catalog entry is greater than the current page's upper bound. This means we're done
+                        // with the current page. Move to the next page and reset the item index.
+                        pageIndex++;
+                        itemIndex = 0;
+                    }
+                }
+            }
+
+            // Now that one of the two input lists is drained, handle the other (undrained) list. 
+
+            // Process the rest of the catalog leaves, if any.
+            if (catalogIndex < sortedCatalog.Count)
+            {
+                // Make sure there is at least one non-full page so that remaining catalog leaves can be pushed there.
+                if (pageInfos.Count == 0
+                    || pageInfos.Last().Count == MaxLeavesPerPage)
+                {
+                    context.IndexInfo.Insert(pageIndex, PageInfo.New());
+                }
+                else
+                {
+                    pageIndex = pageInfos.Count - 1;
+                }
+
+                // Push the remaining catalog leaves into the last page.
+                while (catalogIndex < sortedCatalog.Count)
+                {
+                    var catalog = sortedCatalog[catalogIndex];
+                    itemIndex = await MergeEntryIntoPageAsync(context, catalog, pageIndex, itemIndex);
+                    catalogIndex++;
+                }
+                RemovePageAtIfEmpty(context, pageIndex);
+            }
+
+            // Process the rest of the registration pages, if any, by ensuring the remaining pages are not too large.
+            while (pageIndex < pageInfos.Count)
+            {
+                await EnsureValidPageSizeAtAsync(context, pageIndex);
+                pageIndex++;
+            }
+        }
+
+        private async Task<int> MergeEntryIntoPageAsync(
+            Context context,
+            CatalogCommitItem entry,
+            int pageIndex,
+            int itemIndex)
+        {
+            var pageInfo = context.IndexInfo.Items[pageIndex];
+            var items = await pageInfo.GetLeafInfosAsync();
+
+            while (itemIndex < items.Count)
+            {
+                if (entry.PackageIdentity.Version > items[itemIndex].Version)
+                {
+                    // The current position in the item list is too low for the catalog version. Keep looking.
+                    itemIndex++;
+                }
+                else if (entry.PackageIdentity.Version == items[itemIndex].Version)
+                {
+                    if (entry.IsPackageDelete)
+                    {
+                        // Remove the registration leaf item. The current catalog commit item represents a
+                        // delete for this version.
+                        _logger.LogInformation(
```

Comment:
Does this logging context have the package id? If not, do you think it's worth adding the package id to these logs for debugging purposes?

Reply:
It's part of the logging context. I am hoping to lean on that entirely. Logging the package IDs in every log line would be painful.